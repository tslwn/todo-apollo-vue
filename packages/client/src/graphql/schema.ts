export interface Todo {
  __typename: 'Todo';
  id: number;
  text: string;
  isComplete: boolean;
  isArchived: boolean;
  createdAt?: string;
  updatedAt?: string;
}

export enum Sort {
  Asc = 'ASC',
  Desc = 'DESC',
}

export type TodosFilterInput = Partial<Pick<Todo, 'isComplete' | 'isArchived'>>;

export type TodosOrderByInputKey = keyof Pick<Todo, 'createdAt'>;

export type TodosOrderByInput = {
  [K in TodosOrderByInputKey]?: Sort;
};

export interface TodosInput {
  filter?: TodosFilterInput;
  orderBy?: TodosOrderByInput;
}

export type TodoInput = Pick<Todo, 'id'>;

export interface Query {
  __typename: 'Query';
}

export interface TodosResponse extends Query {
  todos: Todo[];
}

export type AddTodoInput = Pick<Todo, 'text'>;

export type ChangeTodoTextInput = Pick<Todo, 'id' | 'text'>;

export type ChangeTodoIsCompleteInput = Pick<Todo, 'id' | 'isComplete'>;

export type ChangeTodoIsArchivedInput = Pick<Todo, 'id' | 'isArchived'>;

interface TodoUpdateResponseGeneric {
  __typename: 'TodoUpdateResponse';
  success: boolean;
  message?: string;
}

interface TodoUpdateResponseSuccess extends TodoUpdateResponseGeneric {
  success: true;
  todo: Pick<Todo, '__typename' | 'id' | 'text' | 'isComplete' | 'isArchived'>;
}

interface TodoUpdateResponseFailure extends TodoUpdateResponseGeneric {
  success: false;
}

export type TodoUpdateResponse =
  | TodoUpdateResponseSuccess
  | TodoUpdateResponseFailure;

export interface Mutation {
  __typename: 'Mutation';
}

export interface AddTodoResponse extends Mutation {
  addTodo: TodoUpdateResponse;
}

export interface ChangeTodoTextResponse extends Mutation {
  changeTodoText: TodoUpdateResponse;
}

export interface ChangeTodoIsCompleteResponse extends Mutation {
  changeTodoIsComplete: TodoUpdateResponse;
}

export interface ChangeTodoIsArchivedResponse extends Mutation {
  changeTodoIsArchived: TodoUpdateResponse;
}
