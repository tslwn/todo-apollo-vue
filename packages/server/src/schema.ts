import { gql } from 'apollo-server';
import { Todo, TodoAttributes } from 'types/todo.types';

const schema = gql`
  type Todo {
    id: ID!
    text: String!
    isComplete: Boolean!
    isArchived: Boolean!
    createdAt: String!
    updatedAt: String!
  }

  enum Sort {
    ASC
    DESC
  }

  input TodoFilterInput {
    isComplete: Boolean
    isArchived: Boolean
  }

  input TodoOrderByInput {
    createdAt: Sort
  }

  type Query {
    todos(filter: TodoFilterInput, orderBy: TodoOrderByInput): [Todo]!
    todo(id: ID!): Todo
  }

  type TodoUpdateResponse {
    success: Boolean!
    message: String
    todo: Todo
  }

  type Mutation {
    addTodo(text: String!): TodoUpdateResponse!
    changeTodoText(id: ID!, text: String!): TodoUpdateResponse!
    changeTodoIsComplete(id: ID!, isComplete: Boolean!): TodoUpdateResponse!
    changeTodoIsArchived(id: ID!, isArchived: Boolean!): TodoUpdateResponse!
  }
`;

export enum Sort {
  ASC,
  DESC
}

export type TodosFilterInput = Partial<Pick<Todo, 'isComplete' | 'isArchived'>>;

type TodosOrderByInputKey = keyof Pick<Todo, 'createdAt'>;

export type TodosOrderByInput = {
  [index: string]: { [K in TodosOrderByInputKey]?: Sort };
};

export interface TodosInput {
  filter?: TodosFilterInput;
  orderBy?: TodosOrderByInput;
}

export type TodoInput = Pick<Todo, 'id'>;

export type AddTodoInput = Pick<Todo, 'text'>;

export type ChangeTodoTextInput = Pick<Todo, 'id' | 'text'>;

export type ChangeTodoIsCompleteInput = Pick<Todo, 'id' | 'isComplete'>;

export type ChangeTodoIsArchivedInput = Pick<Todo, 'id' | 'isArchived'>;

export interface TodoUpdateResponse {
  success: boolean;
  message?: string;
  todo?: Todo;
}

export default schema;
