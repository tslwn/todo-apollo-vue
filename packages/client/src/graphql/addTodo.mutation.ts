import gql from 'graphql-tag';
import { MutationUpdaterFn } from 'apollo-boost';
import { AddTodoResponse, TodosResponse } from './schema';
import { todosQuery, TODOS_VARIABLES } from './todos.query';

export const addTodoMutation = gql`
  mutation($text: String!) {
    addTodo(text: $text) {
      success
      message
      todo {
        id
        text
        isComplete
        isArchived
      }
    }
  }
`;

export const addTodoUpdate: MutationUpdaterFn<AddTodoResponse> = (
  store,
  mutationResult,
) => {
  const { data } = mutationResult;
  if (!data) {
    throw new Error('No mutation result');
  }

  const { message, success, todo } = data.addTodo;
  if (!success) {
    throw new Error(message);
  }

  const queryResponse = store.readQuery<TodosResponse>({
    query: todosQuery,
    variables: TODOS_VARIABLES,
  });
  if (queryResponse === null) {
    throw new Error('No query result');
  }

  // add todo to cache
  const todos =
    TODOS_VARIABLES?.orderBy?.createdAt === 'DESC'
      ? [todo, ...queryResponse.todos]
      : [...queryResponse.todos, todo];

  store.writeQuery({
    data: {
      todos,
    },
    query: todosQuery,
    variables: TODOS_VARIABLES,
  });
};
