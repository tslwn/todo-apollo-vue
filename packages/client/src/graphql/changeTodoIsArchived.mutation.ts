import gql from 'graphql-tag';
import { MutationUpdaterFn } from 'apollo-boost';
import { Todo, ChangeTodoIsArchivedResponse, TodosResponse } from './schema';
import { todosQuery, TODOS_VARIABLES } from './todos.query';

export const changeTodoIsArchivedMutation = gql`
  mutation($id: ID!, $isArchived: Boolean!) {
    changeTodoIsArchived(id: $id, isArchived: $isArchived) {
      success
      message
      todo {
        id
        text
        isComplete
        isArchived
      }
    }
  }
`;

export const changeTodoIsArchivedUpdate: MutationUpdaterFn<ChangeTodoIsArchivedResponse> = (
  store,
  mutationResult,
) => {
  const { data } = mutationResult;
  if (data == null) {
    throw new Error('No mutation result');
  } else if (data.changeTodoIsArchived.success === false) {
    throw new Error(data.changeTodoIsArchived.message);
  } else {
    const todo = data.changeTodoIsArchived.todo;

    // get todos before update
    const queryResponse = store.readQuery<TodosResponse>({
      query: todosQuery,
      variables: TODOS_VARIABLES,
    });
    if (queryResponse === null) {
      throw new Error('No query result');
    }

    // remove todo from query cache if not included by filter
    const todoInFilter =
      TODOS_VARIABLES?.filter?.isArchived === undefined ||
      TODOS_VARIABLES.filter.isArchived === todo.isArchived;

    if (!todoInFilter) {
      store.writeQuery({
        data: {
          todos: queryResponse.todos.filter(
            (item: Todo) => item.id !== todo?.id,
          ),
        },
        query: todosQuery,
        variables: TODOS_VARIABLES,
      });
    }
  }
};

/**
 * TODO: distinguish between 'attributes' and 'creation attributes' following
 * server package `src/types/todo.types`
 */
export const changeTodoIsArchivedOptimisticResponse = ({
  id,
  text,
  isComplete,
  isArchived,
}: Pick<
  Todo,
  'id' | 'text' | 'isComplete' | 'isArchived'
>): ChangeTodoIsArchivedResponse => ({
  __typename: 'Mutation',
  changeTodoIsArchived: {
    message: 'Todo isArchived changed successfully',
    success: true,
    todo: {
      __typename: 'Todo',
      id,
      text,
      isComplete,
      isArchived: !isArchived,
    },
    __typename: 'TodoUpdateResponse',
  },
});
