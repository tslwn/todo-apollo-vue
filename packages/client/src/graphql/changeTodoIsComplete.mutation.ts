import gql from 'graphql-tag';

export default gql`
  mutation($id: ID!, $isComplete: Boolean!) {
    changeTodoIsComplete(id: $id, isComplete: $isComplete) {
      success
      message
      todo {
        id
        text
        isComplete
        isArchived
      }
    }
  }
`;
