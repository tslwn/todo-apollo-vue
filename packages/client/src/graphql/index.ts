import addTodoMutation from './addTodo.mutation';
import changeTodoIsArchivedMutation from './changeTodoIsArchived.mutation';
import changeTodoIsCompleteMutation from './changeTodoIsComplete.mutation';
import todosQuery from './todos.query';

export {
  addTodoMutation,
  changeTodoIsArchivedMutation,
  changeTodoIsCompleteMutation,
  todosQuery,
};
