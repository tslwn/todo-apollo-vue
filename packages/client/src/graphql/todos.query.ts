import gql from 'graphql-tag';

export default gql`
  query($filter: TodoFilterInput, $orderBy: TodoOrderByInput) {
    todos(filter: $filter, orderBy: $orderBy) {
      id
      text
      isComplete
      isArchived
    }
  }
`;
