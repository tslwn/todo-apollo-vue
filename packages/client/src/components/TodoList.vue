<template>
  <div>
    <v-slide-x-transition group tag="v-list">
      <todo
        v-for="todo in todosWithHash"
        v-bind:key="todo.hash"
        v-bind:todo="todo"
      ></todo>
    </v-slide-x-transition>
  </div>
</template>

<script lang="ts">
import Vue from 'vue';
import { todosQuery, TODOS_VARIABLES } from '../graphql/todos.query';
import Todo from './Todo.vue';
import { Todo as ITodo } from '../graphql/schema';

interface TodoWithHash extends ITodo {
  hash: string;
}

export default Vue.extend({
  name: 'todo-list',
  components: {
    Todo,
  },
  data: () => ({ todos: [] as ITodo[] }),
  computed: {
    todosWithHash(): TodoWithHash[] {
      return this.todos.map((todo) => ({ hash: btoa(todo.text), ...todo }));
    },
  },
  apollo: {
    todos: {
      query: todosQuery,
      variables: TODOS_VARIABLES,
    },
  },
});
</script>

<style scoped></style>
