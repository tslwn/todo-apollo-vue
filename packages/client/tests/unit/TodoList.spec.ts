import { graphql } from 'graphql';
import { addMocksToSchema } from '@graphql-tools/mock';
import { v4 as uuidv4 } from 'uuid';
import Vue from 'vue';
import Vuetify from 'vuetify';
import { mount, createLocalVue } from '@vue/test-utils';
import TodoList from '../../src/components/TodoList.vue';
import schema from '../mockSchema';

Vue.use(Vuetify);

const localVue = createLocalVue();

describe('TodoList.vue', () => {
  let vuetify: typeof Vuetify;

  beforeEach(() => {
    vuetify = new Vuetify();
    addMocksToSchema({ schema, preserveResolvers: true });
  });

  /**
   * Snapshot tests. Note that we are mocking the todos query data directly
   * rather than the schema response.
   */

  const todos = [
    {
      id: uuidv4(),
      text: 'Add unit tests',
      isComplete: false,
      isArchived: false,
    },
    {
      id: uuidv4(),
      text: 'Add integration tests',
      isComplete: true,
      isArchived: true,
    },
  ];

  it('renders correctly', () => {
    const wrapper = mount(TodoList, { localVue, vuetify });
    wrapper.setData({ todos });
    expect(wrapper).toMatchSnapshot();
  });

  /**
   * Mock schema tests. Here we are mocking the schema response.
   */

  it('renders correctly with mock schema', async () => {
    const query = `query($filter: TodoFilterInput, $orderBy: TodoOrderByInput) {
      todos(filter: $filter, orderBy: $orderBy) {
        id
        text
        isComplete
        isArchived
      }
    }`;
    const wrapper = mount(TodoList, { localVue, vuetify });
    const result = await graphql(schema, query);
    if (result.data == null) {
      throw new Error('No query result');
    }
    wrapper.setData(result.data);
    expect(wrapper).toMatchSnapshot();
  });
});
