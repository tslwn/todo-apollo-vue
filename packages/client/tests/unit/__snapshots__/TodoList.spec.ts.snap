// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList.vue renders correctly 1`] = `
<div>
  <transition-group-stub tag="v-list" name="slide-x-transition"></transition-group-stub>
</div>
`;

exports[`TodoList.vue renders correctly with mock schema 1`] = `
<div>
  <transition-group-stub tag="v-list" name="slide-x-transition"></transition-group-stub>
</div>
`;
